flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/Executing.java

```diff
@@ -129,6 +134,43 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
         executionVertex.markFailed(e);
     }
 
+    @Override
+    public void notifyNewResourcesAvailable() {
+        int availableSlots = context.getAvailableSlots();
+
+        if (availableSlots > 0) {
+            // check if rescaling would increase the parallelism of at least one vertex
+
+            final Optional<? extends VertexAssignment> potentialNewParallelism =
+                    context.getVertexAssignmentWithAllSlots();
```

Comment:
Instead of introducing two new methods one could introduce a method `boolean Context.canScaleUp()` which is implemented by the `DeclarativeScheduler`. The implementation could then call into a `RescalingStrategy` which makes this decision. That way the interaction between the `Executing` state and the `Executing.Context` implementation is a bit easier.

Reply:
Thx for the feedback. I moved the logic into the scheduler.

I assume introducing the RescalingStrategy should be done once we've agreed on an interface in the design document, and then done in a separate PR.

Reply:
Yes, exactly.